Add tests for window.event

miketaylr opened the pull request:
Corresponding DOM PR @ https://github.com/whatwg/dom/pull/407

Corresponding DOM issue @ https://github.com/whatwg/dom/issues/334

<!-- Reviewable:start -->

<!-- Reviewable:end -->

w3c-bots commented:
# Chrome (unstable channel) #
Testing web-platform-tests at revision ab2fa21b2113d36c6e25604b18cc73e0333214cc
Using browser at version 58.0.3004.3 dev
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary><a href="http://w3c-test.org/submissions/4790/dom/events/Event-initEvent.html">/dom/events/Event-initEvent.html</a></summary>


|                               Subtest                               | Results |
|---------------------------------------------------------------------|---------|
|                                                                     | OK      |
| `Properties of initEvent(type, true, false)`                        | PASS    |
| `Calling initEvent during propagation.`                             | PASS    |
| `Properties of initEvent(type, false, true)`                        | PASS    |
| `Calling initEvent must unset the stop immediate propagation flag.` | PASS    |
| `Calling initEvent must unset the stop propagation flag.`           | PASS    |
| `Calling initEvent multiple times (getting type).`                  | PASS    |
| `Calling initEvent multiple times (not getting type).`              | PASS    |
| `Properties of initEvent(type, true, true)`                         | PASS    |
| `Properties of initEvent(type, false, false)`                       | PASS    |
| `Calling initEvent must not have an effect during dispatching.`     | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4790/dom/events/Event-defaultPrevented-after-dispatch.html">/dom/events/Event-defaultPrevented-after-dispatch.html</a></summary>


|                          Subtest                           | Results |
|------------------------------------------------------------|---------|
|                                                            | OK      |
| `Event.defaultPrevented is not reset after dipatchEvent()` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4790/dom/events/EventListener-handleEvent.html">/dom/events/EventListener-handleEvent.html</a></summary>


|            Subtest             | Results |
|--------------------------------|---------|
|                                | OK      |
| `EventListener::handleEvent()` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4790/dom/events/Event-dispatch-other-document.html">/dom/events/Event-dispatch-other-document.html</a></summary>


|                     Subtest                      | Results |
|--------------------------------------------------|---------|
|                                                  | OK      |
| `Custom event on an element in another document` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4790/dom/events/Event-dispatch-click.html">/dom/events/Event-dispatch-click.html</a></summary>


|                             Subtest                             | Results |
|-----------------------------------------------------------------|---------|
|                                                                 | OK      |
| `event state during post-click handling`                        | PASS    |
| `redispatch during post-click handling`                         | PASS    |
| `disabled checkbox still has activation behavior`               | PASS    |
| `pick the first with activation behavior <a href>`              | PASS    |
| `disconnected radio should be checked`                          | PASS    |
| `disconnected form should not submit`                           | FAIL    |
| `basic with wrong event class`                                  | PASS    |
| `disabled checkbox still has activation behavior, part 2`       | PASS    |
| `pick the first with activation behavior <input type=checkbox>` | FAIL    |
| `disconnected checkbox should be checked`                       | PASS    |
| `look at parents when event bubbles`                            | FAIL    |
| `look at parents only when event bubbles`                       | PASS    |
| `basic with click()`                                            | PASS    |
| `basic with dispatchEvent()`                                    | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4790/dom/events/Event-constructors.html">/dom/events/Event-constructors.html</a></summary>


|         Subtest         | Results |
|-------------------------|---------|
|                         | OK      |
| `Event constructors 11` | PASS    |
| `Event constructors 10` | PASS    |
| `Event constructors`    | PASS    |
| `Event constructors 3`  | PASS    |
| `Event constructors 2`  | PASS    |
| `Event constructors 1`  | PASS    |
| `Event constructors 7`  | PASS    |
| `Event constructors 6`  | PASS    |
| `Event constructors 5`  | PASS    |
| `Event constructors 4`  | PASS    |
| `Event constructors 9`  | PASS    |
| `Event constructors 8`  | PASS    |
| `Event constructors 12` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4790/dom/events/Event-dispatch-detached-click.html">/dom/events/Event-dispatch-detached-click.html</a></summary>


|                                  Subtest                                   | Results |
|----------------------------------------------------------------------------|---------|
|                                                                            | OK      |
| `Click event can be dispatched to an element that is not in the document.` | PASS    |
| `Click event on an element not in the document`                            | PASS    |

</details>

w3c-bots commented:
# Firefox (nightly channel) #
Testing web-platform-tests at revision ab2fa21b2113d36c6e25604b18cc73e0333214cc
Using browser at version BuildID 20170123125947; SourceStamp 36486fdc3813ef7943ae5b07b4128866d1938a6c
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary><a href="http://w3c-test.org/submissions/4790/dom/events/Event-initEvent.html">/dom/events/Event-initEvent.html</a></summary>


|                               Subtest                               | Results |
|---------------------------------------------------------------------|---------|
|                                                                     | OK      |
| `Properties of initEvent(type, true, false)`                        | FAIL    |
| `Calling initEvent during propagation.`                             | PASS    |
| `Properties of initEvent(type, false, true)`                        | FAIL    |
| `Calling initEvent must unset the stop immediate propagation flag.` | PASS    |
| `Calling initEvent must unset the stop propagation flag.`           | PASS    |
| `Calling initEvent multiple times (getting type).`                  | PASS    |
| `Calling initEvent multiple times (not getting type).`              | PASS    |
| `Properties of initEvent(type, true, true)`                         | FAIL    |
| `Properties of initEvent(type, false, false)`                       | FAIL    |
| `Calling initEvent must not have an effect during dispatching.`     | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4790/dom/events/Event-defaultPrevented-after-dispatch.html">/dom/events/Event-defaultPrevented-after-dispatch.html</a></summary>


|                          Subtest                           | Results |
|------------------------------------------------------------|---------|
|                                                            | OK      |
| `Event.defaultPrevented is not reset after dipatchEvent()` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4790/dom/events/EventListener-handleEvent.html">/dom/events/EventListener-handleEvent.html</a></summary>


|            Subtest             | Results |
|--------------------------------|---------|
|                                | OK      |
| `EventListener::handleEvent()` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4790/dom/events/Event-dispatch-other-document.html">/dom/events/Event-dispatch-other-document.html</a></summary>


|                     Subtest                      | Results |
|--------------------------------------------------|---------|
|                                                  | OK      |
| `Custom event on an element in another document` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4790/dom/events/Event-dispatch-click.html">/dom/events/Event-dispatch-click.html</a></summary>


|                             Subtest                             | Results |
|-----------------------------------------------------------------|---------|
|                                                                 | OK      |
| `event state during post-click handling`                        | FAIL    |
| `redispatch during post-click handling`                         | FAIL    |
| `disabled checkbox still has activation behavior`               | PASS    |
| `pick the first with activation behavior <a href>`              | FAIL    |
| `disconnected radio should be checked`                          | FAIL    |
| `disconnected form should not submit`                           | PASS    |
| `basic with wrong event class`                                  | PASS    |
| `disabled checkbox still has activation behavior, part 2`       | FAIL    |
| `pick the first with activation behavior <input type=checkbox>` | FAIL    |
| `disconnected checkbox should be checked`                       | FAIL    |
| `look at parents when event bubbles`                            | PASS    |
| `look at parents only when event bubbles`                       | FAIL    |
| `basic with click()`                                            | PASS    |
| `basic with dispatchEvent()`                                    | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4790/dom/events/Event-constructors.html">/dom/events/Event-constructors.html</a></summary>


|         Subtest         | Results |
|-------------------------|---------|
|                         | OK      |
| `Event constructors 11` | PASS    |
| `Event constructors 10` | PASS    |
| `Event constructors`    | PASS    |
| `Event constructors 3`  | FAIL    |
| `Event constructors 2`  | FAIL    |
| `Event constructors 1`  | PASS    |
| `Event constructors 7`  | PASS    |
| `Event constructors 6`  | PASS    |
| `Event constructors 5`  | PASS    |
| `Event constructors 4`  | PASS    |
| `Event constructors 9`  | PASS    |
| `Event constructors 8`  | PASS    |
| `Event constructors 12` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4790/dom/events/Event-dispatch-detached-click.html">/dom/events/Event-dispatch-detached-click.html</a></summary>


|                                  Subtest                                   | Results |
|----------------------------------------------------------------------------|---------|
|                                                                            | OK      |
| `Click event can be dispatched to an element that is not in the document.` | FAIL    |
| `Click event on an element not in the document`                            | PASS    |

</details>

zcorpan commented:
Changes LGTM.

Can you also change /dom/interfaces.html ?

miketaylr commented:
> Can you also change /dom/interfaces.html ?

Ah yeah, will do.

miketaylr commented:
I'm planning on landing `srcElement`, `returnValue` and `window.event` in a single PR @ https://github.com/whatwg/dom/pull/407.

@zcorpan do you think we should land all the tests in a single PR as well, or do you prefer PRs for each?

miketaylr commented:
A single PR would be nice, that way I could prevent merge conflicts w/ https://github.com/w3c/web-platform-tests/pull/4305 -- @cvrebert, do you mind if I pull in your tests into this branch?

cvrebert commented:
Nope

zcorpan commented:
1 PR is OK

w3c-bots commented:
These tests are now available on [w3c-test.org](https://w3c-test.org/submissions/4790)

miketaylr commented:
(not ready for review, just fixing merge conflicts)

w3c-bots commented:
# Build ERRORED

Started: 2017-08-10 21:11:40
Finished: 2017-08-10 22:01:34





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/17238)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/263261821)

Commit:
Add tests for window.event

Refs https://github.com/whatwg/dom/issues/334

miketaylr commented:
@zcorpan, would you mind reviewing the window.event tests? (unless someone else wants to jump in)

zcorpan commented on dom/events/event-global.html:
```diff
@@ -0,0 +1,92 @@
+    assert_equals(window.event.type, "cool", "got expected event from global event during dispatch");
+    target2.dispatchEvent(new Event("dude"));
+    assert_not_equals(window.event.type, "dude", "got expected event from global event after handling a different event handler callback");
+    assert_equals(window.event.type, "cool", "got expected event from global event after handling a different event handler callback");
```
Doesn't the spec PR say that this should be undefined? If window.event needs to be able to change to different events and back, maybe it needs to be a stack (internally)?

miketaylr replied:
Let me re-publish an email exchange I had with @foolip:

> On Jan 26, 2018, at 2:51 AM, Philip Jägenstedt <...> wrote:
> Here's the code in Chromium:
https://cs.chromium.org/chromium/src/third_party/WebKit/Source/bindings/core/v8/custom/V8WindowCustom.cpp
> Looks like this is stored as private property on each window instance, in other words not a stack.
> 
> In https://cs.chromium.org/chromium/src/third_party/WebKit/Source/bindings/core/v8/V8AbstractEventListener.cpp, the old value is saved and then restored after the event listener was invoked.
>
> This looks roughly like step 2.7 of https://dom.spec.whatwg.org/#concept-event-listener-inner-invoke. If there's any way to observe the difference between doing this around each individual invoke, or for the whoe dispatch, that matters. (I can't think of a way though.)

On Thu, Jan 25, 2018 at 1:07 AM Mike Taylor <...> wrote:
>> So here’s my understanding:
>>
>> During event dispatch:
>> * set window.event to the current event (unless we’re in shadow doc land, we leave it as undefined).
>>
>> What I’m not sure how to define is the scenario where it’s possible to dispatch an event from a listener’s callback — so we need to change window.event to the new current event during that dispatch, and then restore it back to the old current event (when we return to the callback from inner invoke).
>>
>> So… should we define this as a stack? Or am I overthinking this?

miketaylr replied:
OK, re-reading that and looking at https://whatpr.org/dom/407.html#ref-for-concept-event-listener-inner-invoke:

If I set *current event* to `event` in between Steps 1 and 2 of inner invoke, and set it to undefined after Step 2, this should work, I think.

Thinking out loud:

When `cool` is dispatched:
*  do the `cool` inner invoke and set *current event* to `cool`

During the `cool` listener's callback, `dude` is dispatched:
*  do the `dude` inner invoke and set *current event* to `dude`

Once that is done, resume step 7 of the `cool` inner invoke and its *current event* (`cool`) is in scope again, right?

zcorpan replied:
Yeah... except instead of setting to undefined afterwards, restore to the saved value (might be undefined or an event).

zcorpan commented on dom/events/event-global.html:
```diff
@@ -0,0 +1,92 @@
+  }));
+
+  shadowNode.dispatchEvent(new Event("test", {composed: true, bubbles: true}));
+}, "window.event is undefined if the target is in a shadow tree, when event is dispatched inside shadow tree");
```
I don't understand how this is explained by the spec PR...

miketaylr replied:
Yeah, I need to square this test with the spec PR. I discovered the difference writing tests last night, which surprised me.

zcorpan reviewed (changes requested):
Is the spec PR implemented? I think the spec is a bit wrong, probably the tests are more correct. But maybe I'm missing something!

miketaylr commented on dom/events/event-global.html:
```diff
@@ -0,0 +1,92 @@
+<div id=log></div>
+<script>
+test(t => {
+  assert_own_property(window, "event");
```
Thanks.

miketaylr commented:
I believe I addressed all review feedback in https://github.com/w3c/web-platform-tests/pull/4790/commits/2a97b2c332857a4fa0fc85dd1f038d145a71aba7.

r? @zcorpan

miketaylr commented:
Split out srcElement and returnValue in https://github.com/w3c/web-platform-tests/pull/10258, so I'm going to remove those commits from this PR.

zcorpan commented:
Sorry for the delay

annevk committed:
fix error in dom/interfaces.html

annevk committed:
align IDL location with spec

annevk commented:
I made some further changes (and rebased again) to avoid conflicting with more recent IDL setup changes. I think this is otherwise still good to go.
